Add math operators and functions to work with multidimensional vectors

mathalex opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

**Changelog category:**
- New Feature

**Changelog entry:**
This fully closes #4509 and even more.

**Detailed description / Documentation draft:**

- Function `tupleHammingDistance` has been refactored. Now the summation is performed within only one variable instead of a column to get the result.

- `operator +`, `tuplePlus`, `vectorSum` — do tuple-wise addition. Arguments: `(Tuple, Tuple)`. Returns: `Tuple`.
- `operator -`, `tupleMinus`, `vectorDifference` — do tuple-wise subtraction. Arguments: `(Tuple, Tuple)`. Returns: `Tuple`.
- `tupleMultiply` — do tuple-wise multiplication (compatibility). Arguments: `(Tuple, Tuple)`. Returns: `Tuple`.
- `tupleDivide` — do tuple-wise division (compatibility). Arguments: `(Tuple, Tuple)`. Returns: `Tuple`.

- `unary operator -`, `tupleNegate` — do tuple-wise negation. Arguments: `(Tuple)`. Returns: `Tuple`.

- `operator *`, `tupleMultiplyByNumber` — multiplies each element of a tuple by a number. Arguments: `(Tuple, Number)`. Returns: `Tuple`.
- `operator /`, `tupleDivideByNumber` — divides each element of a tuple by a number. Arguments: `(Tuple, Number)`. Returns: `Tuple`.

- `operator *`, `dotProduct`, `scalarProduct` — a dot (aka scalar) product of vectors. Arguments: `(Tuple, Tuple)`. Returns: `Number`.

- `L1Norm`, `normL1` — calculates the sum of absolute values of coordinates. Arguments: `(Tuple)`. Returns: `Number`.
- `L2Norm`, `normL2` — calculates the square root of the sum of coordinates squares. Arguments: `(Tuple)`. Returns: `Number`.
- `LinfNorm`, `normLinf` — calculates the maximum absolute value among coordinates. Arguments: `(Tuple)`. Returns: `Number`.
- `LpNorm`, `normLp` — calculates a root of `p`th power of the sum of absolute values of coordinates in `p`th powers. Arguments: `(Tuple, Number)`. Returns: `Number`. ***`LpNorm` should be reviewed very carefully***.

- `L1Distance`, `distanceL1` — finds the distance between two points (as tuples) using 1-norm. Arguments: `(Tuple, Tuple)`. Returns: `Number`.
- `L2Distance`, `distanceL2` — finds the distance between two points (as tuples) using 2-norm. Arguments: `(Tuple, Tuple)`. Returns: `Number`.
- `LinfDistance`, `distanceLinf` — finds the distance between two points (as tuples) using infinity-norm. Arguments: `(Tuple, Tuple)`. Returns: `Number`.
- `LpDistance`, `distanceLp` — finds the distance between two points (as tuples) using p-norm. Arguments: `(Tuple, Tuple, Number)`. Returns: `Number`.

- `L1Normalize`, `normalizeL1` — finds a unit vector of a given vector (tuple) according to 1-norm. Arguments: `(Tuple)`. Returns: `Tuple`.
- `L2Normalize`, `normalizeL2` — finds a unit vector of a given vector (tuple) according to 2-norm. Arguments: `(Tuple)`. Returns: `Tuple`.
- `LinfNormalize`, `normalizeLinf` — finds a unit vector of a given vector (tuple) according to infinity-norm. Arguments: `(Tuple)`. Returns: `Tuple`.
- `LpNormalize`, `normalizeLp` — finds a unit vector of a given vector (tuple) according to p-norm. Arguments: `(Tuple, Number)`. Returns: `Tuple`.

- `cosineDistance` — calculates the cosine of the angle between vectors and subtracts it from one. Arguments: `(Tuple, Tuple)`. Returns: `Number`.

- `max2` — finds the maximum of two numbers (developed for `LinfNorm` function and it is just good to have this function). Arguments: `(Number, Number)`. Returns: `Number`. ***Maybe should be updated as there is a conversion to Float64, as I can understand.***
- `min2`  — finds the minimum of two numbers (compatibility). Arguments: `(Number, Number)`. Returns: `Number`.

<hr>

Examples for each of the queries can be founded in the test related to this pull request.

In `Lp` functions only when `p` is not less than 1 makes sense as it is not a norm in the opposite case. However, there are no restrictions, so the user can pass even a negative number as a parameter.
**UPD**: added restrictions `1 <= p < inf`.

`LxDistance(u, v) := LxNorm(u - v)`, `LxNormalize(u) := u / LxNorm(u)`, `cosineDistance(u, v) := 1 - (u * v) / (L2Norm(u) * L2Norm(v))`.

Operators overloading that can be added:
- (not related to numeric vectors) Overload + (`(String, String) -> String`) to concatenate strings, * (`(String, Integer) -> String`) to concatenate one string multiple times.

mathalex committed:
Operators, refactoring

Added operators, tupleHammingDistance has been refactored

mathalex committed:
Merge branch 'master' into numeric_tuple_functions

mathalex committed:
Compatibility

mathalex committed:
Deleting cerr

mathalex committed:
Style

mathalex committed:
Divide, negate, test

mathalex committed:
Merge branch 'ClickHouse:master' into numeric_tuple_functions

mathalex committed:
Reduced copy-paste, dotProduct

mathalex committed:
L1Norm

mathalex committed:
L2Norm

mathalex committed:
max2, min2, LinfNorm

mathalex committed:
LpNorm

mathalex committed:
Distances, interesting tests

mathalex committed:
Reduced copy-paste

mathalex committed:
Reduced copy-paste

mathalex committed:
Names, tupleOperatorByNumber

mathalex committed:
Changed operator, fix clarity of test

mathalex committed:
Normalize

mathalex committed:
Speed up

mathalex committed:
cosineDistance

mathalex committed:
operators

mathalex committed:
unary operator

vdimir commented on tests/queries/0_stateless/02011_tuple_vector_functions.sql:
```diff
@@ -0,0 +1,58 @@
+SELECT tupleHammingDistance(tuple(1), tuple(1));
```
Lets also add couple of cases with NULL values.

For example following example return unexpected error (it thrown not from your code, but can be realted?)
```sql
SELECT 2 * (1, 1, NULL);
```
```
Received exception from server (version 21.10.1):
Code: 44. DB::Exception: Received from localhost:9000. DB::Exception: ColumnTuple cannot have ColumnConst as its element: While processing 2 * (1, 1, NULL). (ILLEGAL_COLUMN)
```

mathalex replied:
Yes, it should not fail as operation `2 * NULL` is OK. Will be fixed.

mathalex replied:
Now it is fixed, but there is interesting bug or feature: (1, 2) * NULL is NULL, not a tuple of NULLs.

vdimir commented on tests/queries/0_stateless/02011_tuple_vector_functions.sql:
```diff
@@ -0,0 +1,58 @@
+SELECT tupleHammingDistance(tuple(1), tuple(1));
+SELECT tupleHammingDistance((1, 3), (1, 2));
```
Also it's good to test errors for some corner cases, like number of elements doesn't match, empty tuples (is it possible now?), illegal types and so on.

mathalex replied:
As `tupleHammingDistance` was not written by me, there are a lot of tests for this function in the file `01653_tuple_hamming_distance_2.sql`. I can add such corner cases checks with other functions.
To be honest, I do not know how to create an empty tuple (`tuple()` does not work).

mathalex replied:
Added, you can check the description of the ['Corner cases' commit](7eb725a74358d97f1e3a27d4b78d917b13613b92).

vdimir commented on tests/queries/0_stateless/02011_tuple_vector_functions.sql:
```diff
@@ -0,0 +1,58 @@
+SELECT tupleHammingDistance(tuple(1), tuple(1));
+SELECT tupleHammingDistance((1, 3), (1, 2));
+
+SELECT (1, 2) + tupleMultiply((3, 4), (5, 1)) - (6, 3);
```
All tests in file uses contants, it's good to test with non constant column (adding `materialize` or/and using `numbers(n)` table function or table `system.numbers`/`system.numbers_mt`). 
Actually, because we use `useDefaultImplementationForConstants` behavior  is the same. But it can be changed later and couple of more complex cases don't hurt in any case.

mathalex replied:
Will do it.

vdimir commented on src/Functions/vectorFunctions.cpp:
```diff
@@ -0,0 +1,1067 @@
+        {
+            try
+            {
+                ColumnWithTypeAndName left{left_elements.empty() ? nullptr : left_elements[i], left_types[i], {}};
```
Is there real cases where it's empty?

mathalex replied:
Just copypasted from the `tupleHammingDistance`... Actually, do not know (and I agree -- it is strange if there is a real case). I can delete it from all places. If old tests with `tupleHammingDistance` not fail, then it will be possible to conclude that there are no such cases :)

mathalex replied:
Yes, there are!!! `SELECT tupleHammingDistance(materialize((1, 2)), (1, 4));` -- do not know why, but it fails if you leave only the third part of the ternary operator.

vdimir commented on src/Functions/FunctionBinaryArithmetic.h:
```diff
@@ -645,6 +645,64 @@ class FunctionBinaryArithmetic : public IFunction
+            return {};
+
+        if (isNumber(type0) && is_division)
+            throw Exception("Wrong order of arguments for function " + String(name) + ": argument of numeric type cannot be first.",
```
```suggestion
            throw Exception("Wrong order of arguments for function " + String(name) + ": argument of numeric type cannot be first",
```

mathalex replied:
There was a dot in the similar exception in `getFunctionForIntervalArithmetic` (function in this file which was written not by me). I can remove dots in both exceptions.

vdimir replied:
Yes, let's remove it because sometime exception message is wrapped with other one and we will got something like `argument of numeric type cannot be first.: <some additional info>`

vdimir commented on src/Functions/max2.cpp:
```diff
@@ -0,0 +1,22 @@
+namespace
+{
+    struct Max2Name { static constexpr auto name = "max2"; };
+    using FunctionMax2 = FunctionMathBinaryFloat64<BinaryFunctionVectorized<Max2Name, max>>;
```
> Maybe should be updated as there is a conversion to Float64, as I can understand.

Yes it returns Fload64 for all input types.

```
SELECT toTypeName(max2(2, 1))

┌─toTypeName(max2(2, 1))─┐
│ Float64                │
└────────────────────────┘
```

mathalex replied:
Thanks. I think that it is not good to deduce such an exact function to the `Float64` type, but do not know a convenient way to fix it.

vdimir commented on src/Functions/vectorFunctions.cpp:
```diff
@@ -0,0 +1,1067 @@
+        if (tuple_size == 0)
+            return DataTypeUInt8().createColumnConstWithDefaultValue(input_rows_count);
+
+        const auto & p_column = arguments[1];
```
Is there use-cases for non-constnat `p`, how do you think? Also, shold we limit possible values for `p`, like `0 < p < inf` or make it integer? 

```
SELECT LpNorm((3, 1, 4), 0), LpNorm((3, 1, 4), inf);

┌─LpNorm((3, 1, 4), 0)─┬─LpNorm((3, 1, 4), inf)─┐
│                  inf │                      1 │
└──────────────────────┴────────────────────────┘
```

mathalex replied:
I decided to not limit `p`. It definitely should NOT be just integer (I will add a test when `p` is float). Real math sense is when `p >= 1`, but probably somebody will want to use it for other purposes...

vdimir replied:
Let's make it `1 <= p < inf`?

mathalex replied:
Ok.

mathalex replied:
Restrictions added in the 53d7649c33bb43410b4e1ad52b820da4eebb327f.

vdimir commented on src/Functions/vectorFunctions.cpp:
```diff
@@ -0,0 +1,1067 @@
+        }
+
+        auto divide = FunctionFactory::instance().get("divide", context);
+        ColumnWithTypeAndName one{DataTypeUInt8().createColumnConst(input_rows_count, 1)->convertToFullColumnIfConst(),
```
Do we really need `convertToFullColumnIfConst` ?

mathalex replied:
Maybe not. I will delete the conversion.

vdimir commented:
Excellent work!

Some comments, most of them suggests style changes.

vdimir reviewed (commented):
Excellent work!

Some comments, most of them suggests style change.

vdimir commented:
Btw, did you consider different naming variants for `L1/Lp/Linf/Norm/Distance/...` functions?

Maybe `normL1`, `normLp`, `normLInf`/`normLinf`, `distanceL1`, ... and similar is better (not to start name with capital letter)?

mathalex commented on src/Functions/max2.cpp:
```diff
@@ -0,0 +1,22 @@
+#include <Functions/FunctionFactory.h>
+
+template <typename T>
+T max(T a, T b)
```
No, if it can I would of course use it. Exception message: `Value of type '<overloaded function type>' is not implicitly convertible to 'DB::Float64 (*)(DB::Float64, DB::Float64)' (aka 'double (*)(double, double)')`.

mathalex commented on src/Functions/min2.cpp:
```diff
@@ -0,0 +1,22 @@
+#include <Functions/FunctionFactory.h>
+
+template <typename T>
+T min(T a, T b)
```
As in the case of `std::max`.

mathalex commented on src/Functions/vectorFunctions.cpp:
```diff
@@ -0,0 +1,1067 @@
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+static const char PLUS_NAME[] = "plus";
+static const char MINUS_NAME[] = "minus";
+static const char MULTIPLY_NAME[] = "multiply";
+static const char DIVIDE_NAME[] = "divide";
+
+static constexpr char L1_LABEL[] = "1";
+static constexpr char L2_LABEL[] = "2";
+static constexpr char Linf_LABEL[] = "inf";
+static constexpr char Lp_LABEL[] = "p";
```
Exception message: `Pointer to subobject of string literal is not allowed in a template argument` in line with `using FunctionTuplePlus = FunctionTupleOperator<PLUS_NAME>;`.

mathalex commented on src/Functions/vectorFunctions.cpp:
```diff
@@ -0,0 +1,1067 @@
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+static const char PLUS_NAME[] = "plus";
+static const char MINUS_NAME[] = "minus";
+static const char MULTIPLY_NAME[] = "multiply";
+static const char DIVIDE_NAME[] = "divide";
```
No, we cannot as `func_name` is also used in `auto func = FunctionFactory::instance().get(func_name, context);`.

mathalex commented on src/Functions/TupleIFunction.h:
```diff
@@ -0,0 +1,41 @@
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+class TupleIFunction : public IFunction
```
Will be `ITupleFunction`.

mathalex commented on src/Functions/vectorFunctions.cpp:
```diff
@@ -0,0 +1,1067 @@
+    return res;
+}
+
+template <const char * func_name>
```
We cannot use `std::string` and I have used `const char*`, but your approach is really better as we can just pack desirable string into the struct.

mathalex commented on src/Functions/vectorFunctions.cpp:
```diff
@@ -0,0 +1,1067 @@
+        const auto * cur_tuple = checkAndGetDataType<DataTypeTuple>(arguments[0].type.get());
+
+        if (!cur_tuple)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Argument 0 of function {} should be tuples, got {}",
```
There was a typo in the `tupleHammingDistance` (which was not written by me, I just slightly optimized and do not review thoroughly) and that part was included in every other tuple function. Thanks!

mathalex commented:
> Maybe `normL1`, `normLp`, `normLInf`/`normLinf`, `distanceL1`, ... and similar is better (not to start name with capital letter)?

For the purpose of not starting with a capital letter, new names are better. Current names are good in perspective that they can be read straightforward (Lp-norm, not the norm in Lp metric).

mathalex commented on src/Functions/vectorFunctions.cpp:
```diff
@@ -0,0 +1,1067 @@
+        if (tuple_size == 0)
+            return std::make_shared<DataTypeUInt8>();
+
+        const auto & p_column = arguments[1];
```
Do you suggest extracting `p` from the column and the just make a column with `1/p` with type `Float64` in the end? If so, it will inevitably convert the argument to `Float64`, although eventually `pow` is guaranteed to return `Float64` so this is not a problem. I can do it.
Maybe we can just return `Float64` in `getReturnTypeImpl` as `pow` result type is always `Float64`?

mathalex replied:
Simplified in 53d7649c33bb43410b4e1ad52b820da4eebb327f.

mathalex committed:
Style

mathalex committed:
NULL support (check description)

Interesting bug or feature: (1, 2) * NULL is NULL, not tuple of NULLs.

mathalex committed:
Corner cases (check description)

+ for Strings, * for String and Number are not added as it can be implemented soon. LpNorm cannot get Decimal because of the pow function.

mathalex committed:
Tests with non constant columns

mathalex committed:
Add aliases

vdimir commented:
> AST fuzzer (debug)

Error is related to `cosineDistacnce` and `NULL`, see report. 

If it's difficult to fix we can omit support of `Nullable` arguments (but throw readable error). If it's easy to fix let's do it.

mathalex committed:
Constant p

mathalex committed:
Merge branch 'master' into numeric_tuple_functions

mathalex committed:
Small update

mathalex committed:
Cosine null fix

mathalex committed:
Style

qieqieplus commented:
Any plan for supporting `Array` columns?
I know some functions can be implemented with functions like `ArrayMap` and `ArraySum`, but specialized functions would have much better performance.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=27933) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=27933) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=27933) it.</sub>

mathalex committed:
Merge branch 'ClickHouse:master' into numeric_tuple_functions

vdimir committed:
Do not add extra parentheses for tuple negate unary operator

vdimir commented:
Fuzzer failures on [`192633c`](https://github.com/ClickHouse/ClickHouse/pull/27933/commits/192633cb9ba226a4ed55459e0f73b5569cb1b563)

> AST Fuzzer | failure | Found error: The query formatting is broken

Looks like some issues with formatting [`4acd8f3`](https://github.com/ClickHouse/ClickHouse/pull/27933/commits/4acd8f314737f72892b5063af6b3b7ceedb6ea2b) should help

gyuton commented:
Internal documentation ticket: DOCSUP-16593.
